File: python/02631_IntrProg/week005/thermoEquilibrium.py
import numpy as np
 
def thermoEquilibrium(N, r):
    # Initial number of particles in left and right chamber
    NLeft = N
    NRight = 0
 
    # Time step
    t = 0
 
    # Number of random numbers
    NRand = np.size(r)
 
    # Repeat until chambers have equal number of particles
    while (NLeft != NRight):
        # Increment time
        t = t + 1
         
        # Stop if there are no more random numbers
        if (t > NRand):
            t = 0
            break
     
        # Probability of moving a particle from left to right
        pLR = NLeft / N
     
        # Randomly move a particle
        if (r[t-1] <= pLR):
            # Move one particle from left to right
            NLeft = NLeft - 1
            NRight = NRight + 1
        else:
            # Move one particle from right to left
            NLeft = NLeft + 1
            NRight = NRight - 1
 
    return t

File: MATLAB/02631_IntrProg/week005/thermoEquilibrium.m
function t = thermoEquilibrium(N, r)
% INPUT
%   N : Initial number of gas particles in the left chamber (even integer number)
%   r : Sequence of random numbers (vector of decimal numbers between 0 and 1)
% OUTPUT
%   t : Time-steps used to reach equilibrium (integer number) Return zero if equilibrium is not reached.
Nl = N;
Nr = 0;
t = 0;
while Nl~=Nr
    t = t + 1;
    if t > length(r)
        t = 0;
        break;
    end
    pLR = Nl/N;
    if r(t) <= pLR
        Nr = Nr + 1;
        Nl = Nl - 1;
    else
        Nr = Nr - 1;
        Nl = Nl + 1;
    end
end
    